File: utils/i18n.js
const fs = require('fs');
const path = require('path');

let translations = {};
let currentLang = 'en'; // This is the default language
const DEFAULT_LANG = 'en';

/*** Loads the translation file from the project root.*/
function loadTranslations() {
    try {
        // Build a safe path to the file
        const resolvedPath = path.resolve(__dirname, '../translations.json');
        
        // Read the file's content
        const fileContent = fs.readFileSync(resolvedPath, 'utf-8');
        
        // Convert the text content into a JavaScript object
        translations = JSON.parse(fileContent);

    } catch (error) {
        // If the file is missing or broken, print a warning
        console.warn(`[i18n] WARNING: Could not load translations.json. Error: ${error.message}`);
    }
}

/**
 * Sets the active language for the session.
 */
function setLanguage(langCode) {
    // Check if the language (e.g., "es") exists in our file
    if (langCode && translations[langCode]) {
        currentLang = langCode;
    } else {
        // If not, just use the default (English)
        currentLang = DEFAULT_LANG;
    }
}

/**
 * Gets the translated string for a given key.
 */
function t(key, variables = {}) {
    /**
     * Translates a key, replacing placeholders with variables.
     * e.g., t('greeting', { name: 'User' })
     */
    
    // 1. Try to get the message in the current language
    let message = translations[currentLang]?.[key];
    
    // 2. If not found, fall back to default language (English)
    if (message === undefined) {
        message = translations[DEFAULT_LANG]?.[key];
    }
    
    // 3. If still not found, return the key as a clear warning
    if (message === undefined) {
        return `MISSING_KEY: ${key}`;
    }

    // --- THIS IS THE NEW PART ---
    // 4. Loop through all variables and replace placeholders
    // It will find "{command}" and replace it with the value of variables.command
    for (const varName in variables) {
        const placeholder = `{${varName}}`;
        // Use a loop to replace all instances, in case it appears more than once
        while(message.includes(placeholder)) {
            message = message.replace(placeholder, variables[varName]);
        }
    }
    
    return message;
}

// --- Run the loading function as soon as this file is loaded ---
loadTranslations();

// --- Make the 'setLanguage' and 't' functions usable by other files ---
module.exports = {
    setLanguage,
    t
};

File: python_ver/i18n.py
import json
import os
import sys

# These are our "global" variables, just like in the JS file.
_translations = {}
_current_lang = 'en'  # This is the default language
_DEFAULT_LANG = 'en'

def load_translations():
    """Loads the translation file from the project root."""
    
    # This 'global' keyword lets us modify the _translations variable
    global _translations
    
    try:
        # Get the absolute path to *this* file (i18n.py)
        base_dir = os.path.dirname(os.path.abspath(__file__))
        
        # Build a safe path from this file up to the root
        file_path = os.path.join(base_dir, '..', 'translations.json')
        
        # Open the file for reading with 'utf-8' encoding
        with open(file_path, 'r', encoding='utf-8') as f:
            _translations = json.load(f)
            
    except Exception as e:
        # If anything goes wrong (file not found, bad JSON),
        print(f"[i18n] WARNING: Could not load translations.json. Error: {e}", file=sys.stderr)

def set_language(lang_code: str):
    """Sets the active language for the session."""
    
    global _current_lang
    
    # Check if the lang_code is valid AND exists in our dictionary
    if lang_code and lang_code in _translations:
        _current_lang = lang_code
    else:
        _current_lang = _DEFAULT_LANG

def t(key: str, variables: dict = None) -> str:
    """
    Gets the translated string, replacing placeholders with variables.
    e.g., t('greeting', {'name': 'User'})
    """
    
    if variables is None:
        variables = {}

    # 1. Try to find the key in the current language
    message = _translations.get(_current_lang, {}).get(key)
    
    # 2. If not found, fall back to English
    if message is None:
        message = _translations.get(_DEFAULT_LANG, {}).get(key)
        
    # 3. If still not found, return a warning
    if message is None:
        return f"MISSING_KEY: {key}"
        
    # --- THIS IS THE NEW PART ---
    # 4. Loop through all variables and replace placeholders
    # It will find "{command}" and replace it with the value of variables['command']
    for var_name, value in variables.items():
        message = message.replace(f"{{{var_name}}}", str(value))
    
    return message


load_translations()